### PR TITLE
[SPARK-41971][SQL][PYTHON] Add a config for pandas conversion how to handle struct types

### PR DESCRIPTION
**Pull request by ueshin**
### What changes were proposed in this pull request?

Adds a config for pandas conversion how to handle struct types.

- `spark.sql.execution.pandas.structHandlingMode` (default: `"legacy"`)

The conversion mode of struct type when creating pandas DataFrame.

#### When `"legacy"`, the behavior is the same as before, except that with Arrow and Spark Connect will raise a more readable exception when there are duplicated nested field names.

```py
>>> spark.sql("values (1, struct(1 as a, 2 as a)) as t(x, y)").toPandas()
Traceback (most recent call last):
...
pyspark.errors.exceptions.connect.UnsupportedOperationException: [DUPLICATED_FIELD_NAME_IN_ARROW_STRUCT] Duplicated field names in Arrow Struct are not allowed, got [a, a].
```

#### When `"row"`, convert to Row object regardless of Arrow optimization.

```py
>>> spark.conf.set('spark.sql.execution.pandas.structHandlingMode', 'row')
>>> spark.conf.set('spark.sql.execution.arrow.pyspark.enabled', False)
>>> spark.sql("values (1, struct(1 as a, 2 as b)) as t(x, y)").toPandas()
   x       y
0  1  (1, 2)
>>> spark.sql("values (1, struct(1 as a, 2 as a)) as t(x, y)").toPandas()
   x       y
0  1  (1, 2)
>>> spark.conf.set('spark.sql.execution.arrow.pyspark.enabled', True)
>>> spark.sql("values (1, struct(1 as a, 2 as b)) as t(x, y)").toPandas()
   x       y
0  1  (1, 2)
>>> spark.sql("values (1, struct(1 as a, 2 as a)) as t(x, y)").toPandas()
   x       y
0  1  (1, 2)
```

#### When `"dict"`, convert to dict and use suffixed key names, e.g., `a_0`, `a_1`, if there are duplicated nested field names, regardless of Arrow optimization.

```py
>>> spark.conf.set('spark.sql.execution.pandas.structHandlingMode', 'dict')
>>> spark.conf.set('spark.sql.execution.arrow.pyspark.enabled', False)
>>> spark.sql("values (1, struct(1 as a, 2 as b)) as t(x, y)").toPandas()
   x                 y
0  1  {'a': 1, 'b': 2}
>>> spark.sql("values (1, struct(1 as a, 2 as a)) as t(x, y)").toPandas()
   x                     y
0  1  {'a_0': 1, 'a_1': 2}
>>> spark.conf.set('spark.sql.execution.arrow.pyspark.enabled', True)
>>> spark.sql("values (1, struct(1 as a, 2 as b)) as t(x, y)").toPandas()
   x                 y
0  1  {'a': 1, 'b': 2}
>>> spark.sql("values (1, struct(1 as a, 2 as a)) as t(x, y)").toPandas()
   x                     y
0  1  {'a_0': 1, 'a_1': 2}
```

### Why are the changes needed?

Currently there are three behaviors when `df.toPandas()` with nested struct types:

- vanilla PySpark with Arrow optimization disabled

```py
>>> spark.conf.set('spark.sql.execution.arrow.pyspark.enabled', False)
>>> spark.sql("values (1, struct(1 as a, 2 as b)) as t(x, y)").toPandas()
   x       y
0  1  (1, 2)
```

using `Row` object for struct types.

It can use duplicated field names.

```py
>>> spark.sql("values (1, struct(1 as a, 2 as a)) as t(x, y)").toPandas()
   x       y
0  1  (1, 2)
```

- vanilla PySpark with Arrow optimization enabled

```py
>>> spark.conf.set('spark.sql.execution.arrow.pyspark.enabled', True)
>>> spark.sql("values (1, struct(1 as a, 2 as b)) as t(x, y)").toPandas()
   x                 y
0  1  {'a': 1, 'b': 2}
```

using `dict` for struct types.

It raises an Exception when there are duplicated nested field names:

```py
>>> spark.sql("values (1, struct(1 as a, 2 as a)) as t(x, y)").toPandas()
Traceback (most recent call last):
...
pyarrow.lib.ArrowInvalid: Ran out of field metadata, likely malformed
```

- Spark Connect

```py
>>> spark.sql("values (1, struct(1 as a, 2 as b)) as t(x, y)").toPandas()
   x                 y
0  1  {'a': 1, 'b': 2}
```

using `dict` for struct types.

If there are duplicated nested field names, the duplicated keys are suffixed:

```py
>>> spark.sql("values (1, struct(1 as a, 2 as a)) as t(x, y)").toPandas()
   x                     y
0  1  {'a_0': 1, 'a_1': 2}
```

### Does this PR introduce _any_ user-facing change?

Users will be able to configure the behavior.

### How was this patch tested?

Modified the related tests.

### CODE REVIEW

**Comment by ueshin** [comment]
I'll move the conversion function to `pyspark/sql/pandas/types.py`.

**Comment by xinrong-meng** [comment]
Shall we add `.. versionchanged::` to `toPandas`?

File: python/pyspark/sql/pandas/types.py [hunk 1]
```diff
@@ -462,3 +467,233 @@ def _convert_dict_to_map_items(s: "PandasSeriesLike") -> "PandasSeriesLike":
...
+        return np.dtype("datetime64[ns]")
+    elif type(dt) == TimestampNTZType:
+        return np.dtype("datetime64[ns]")
+    elif type(dt) == DayTimeIntervalType:
+        return np.dtype("timedelta64[ns]")
+    else:
+        return None
+
+
+def _create_converter_to_pandas(
```
Comment (xinrong-meng) [inline python/pyspark/sql/pandas/types.py:504]: I am wondering if PySpark/Pandas data type mappings are ever documented publicly.

**Review by xinrong-meng — APPROVED** [review]

**Comment by xinrong-meng** [comment]
Merged to master, thank you!

**Pull request closed without merging**